### ISSUE TITLE
可否增加如同魔影工坊等配音软件那样的，停顿，延长、语速控制等功能！万分期待

### ISSUE THREAD

**Issue by maxadc** [RVC-Boss/GPT-SoVITS#472]

**Comment by RVC-Boss** [comment]
语速调整会有的

**Comment by hjj-lmx** [comment]
> 语速调整会有的

自定义停顿，现在有方法弄吗？例如在某个地方停顿几秒

**Comment by hjj-lmx** [comment]
> > > 语速调整会有的
> > 
> > 
> > 自定义停顿，现在有方法弄吗？例如在某个地方停顿几秒
> 
> 中文我记得好像有吧，有几个特殊符号是可以转为停顿的音素。不过英文没有。。

请问一下在哪个地方

**Comment by Chi8wah** [comment]
> > > > 语速调整会有的
> > > 
> > > 
> > > 自定义停顿，现在有方法弄吗？例如在某个地方停顿几秒
> > 
> > 
> > 中文我记得好像有吧，有几个特殊符号是可以转为停顿的音素。不过英文没有。。
> 
> 请问一下在哪个地方

`/root/workspace/GPT-SoVITS/GPT_SoVITS/text/cleaner.py` 的 `special` 里有俩对应 SP2 SP3，我记得之前试是有用的，不知道现在还有没有用

**Comment by hjj-lmx** [comment]
> SP2

special = [
    # ("%", "zh", "SP"),
    ("￥", "zh", "SP2"),
    ("^", "zh", "SP3"),
    # ('@', 'zh', "SP4")#不搞鬼畜了，和第二版保持一致吧
]

这是什么意思，我需要在指定地方停顿指定的秒数，例如：测试文本[停顿2s]生成语音

**Comment by Chi8wah** [comment]
> > SP2
> 
> special = [ # ("%", "zh", "SP"), ("￥", "zh", "SP2"), ("^", "zh", "SP3"), # ('@', 'zh', "SP4")#不搞鬼畜了，和第二版保持一致吧 ]
> 
> 这是什么意思，我需要在指定地方停顿指定的秒数，例如：测试文本[停顿2s]生成语音

就是在文本中间插入`￥`、`^` 来造成停顿呀，不过应该不能控制停顿几秒，不过应该输入 `^` 会比输入 `￥` 多停顿一会吧

**Comment by Chi8wah** [comment]
> > > > SP2
> > > 
> > > 
> > > special = [ # ("%", "zh", "SP"), ("￥", "zh", "SP2"), ("^", "zh", "SP3"), # ('@', 'zh', "SP4")#不搞鬼畜了，和第二版保持一致吧 ]
> > > 这是什么意思，我需要在指定地方停顿指定的秒数，例如：测试文本[停顿2s]生成语音
> > 
> > 
> > 就是在文本中间插入`￥`、`^` 来造成停顿呀，不过应该不能控制停顿几秒，不过应该输入 `^` 会比输入 `￥` 多停顿一会吧
> 
> 他这个是怎么设置的停顿，不知道一个`￥`会停顿多久，如果知道就可以按插入多少个来计算了

你就在tts那里输入 `测试文本 ￥ 生成语音` 这样试一下呗

**Comment by hjj-lmx** [comment]
> > > > > SP2
> > > > 
> > > > 
> > > > special = [ # ("%", "zh", "SP"), ("￥", "zh", "SP2"), ("^", "zh", "SP3"), # ('@', 'zh', "SP4")#不搞鬼畜了，和第二版保持一致吧 ]
> > > > 这是什么意思，我需要在指定地方停顿指定的秒数，例如：测试文本[停顿2s]生成语音
> > > 
> > > 
> > > 就是在文本中间插入`￥`、`^` 来造成停顿呀，不过应该不能控制停顿几秒，不过应该输入 `^` 会比输入 `￥` 多停顿一会吧
> > 
> > 
> > 他这个是怎么设置的停顿，不知道一个`￥`会停顿多久，如果知道就可以按插入多少个来计算了
> 
> 你就在tts那里输入 `测试文本 ￥ 生成语音` 这样试一下呗，他这个大概就是在 tts 的时候替换成这种 SP2 SP3 的特殊音素输入给大模型

好像没啥用，他只是把特殊的替换成“，”了

**Comment by Chi8wah** [comment]
> > > > > > SP2
> > > > > 
> > > > > 
> > > > > special = [ # ("%", "zh", "SP"), ("￥", "zh", "SP2"), ("^", "zh", "SP3"), # ('@', 'zh', "SP4")#不搞鬼畜了，和第二版保持一致吧 ]
> > > > > 这是什么意思，我需要在指定地方停顿指定的秒数，例如：测试文本[停顿2s]生成语音
> > > > 
> > > > 
> > > > 就是在文本中间插入`￥`、`^` 来造成停顿呀，不过应该不能控制停顿几秒，不过应该输入 `^` 会比输入 `￥` 多停顿一会吧
> > > 
> > > 
> > > 他这个是怎么设置的停顿，不知道一个`￥`会停顿多久，如果知道就可以按插入多少个来计算了
> > 
> > 
> > 你就在tts那里输入 `测试文本 ￥ 生成语音` 这样试一下呗，他这个大概就是在 tts 的时候替换成这种 SP2 SP3 的特殊音素输入给大模型
> 
> 好像没啥用，他只是把特殊的替换成“，”了

你看的是处理后的文本？你看下返回的音素呢，他不是会在 clean_special 里返回静音音素吗？

**Comment by originwufish** [comment]
def get_tts_wav(ref_wav_path, prompt_text, prompt_language, text, text_languageuage, top_k=15, top_p=0.6, temperature=0.6, speed=1, inp_refs=None, spk="default"):
    infer_sovits = speaker_list[spk].sovits
    vq_model = infer_sovits.vq_model
    hps = infer_sovits.hps

    infer_gpt = speaker_list[spk].gpt
    t2s_model = infer_gpt.t2s_model
    max_sec = infer_gpt.max_sec

    t0 = ttime()
    prompt_text = prompt_text.strip("\n")
    if (prompt_text[-1] not in splits): prompt_text += "。" if prompt_language != "en" else "."
    prompt_language, text = prompt_language, text.strip("\n")
    dtype = torch.float16 if is_half == True else torch.float32
    zero_wav = np.zeros(int(hps.data.sampling_rate * 0.3), dtype=np.float16 if is_half == True else np.float32)
    with torch.no_grad():
        wav16k, sr = librosa.load(ref_wav_path, sr=16000)
        wav16k = torch.from_numpy(wav16k)
        zero_wav_torch = torch.from_numpy(zero_wav)
        if (is_half == True):
            wav16k = wav16k.half().to(device)
            zero_wav_torch = zero_wav_torch.half().to(device)
        else:
            wav16k = wav16k.to(device)
            zero_wav_torch = zero_wav_torch.to(device)
        wav16k = torch.cat([wav16k, zero_wav_torch])
        ssl_content = ssl_model.model(wav16k.unsqueeze(0))["last_hidden_state"].transpose(1, 2)  # .float()
        codes = vq_model.extract_latent(ssl_content)
        prompt_semantic = codes[0, 0]
        prompt = prompt_semantic.unsqueeze(0).to(device)

        refers=[]
        if(inp_refs):
            for path in inp_refs:
                try:
                    refer = get_spepc(hps, path).to(dtype).to(device)
                    refers.append(refer)
                except Exception as e:
                    logger.error(e)
        if(len(refers)==0):
            refers = [get_spepc(hps, ref_wav_path).to(dtype).to(device)]

    t1 = ttime()
    version = vq_model.version
    os.environ['version'] = version
    prompt_language = dict_language[prompt_language.lower()]
    text_languageuage = dict_language[text_languageuage.lower()]
    phones1, bert1, norm_text1 = get_phones_and_bert(prompt_text, prompt_language, version)
    texts = text.split("\n")
    audio_bytes = BytesIO()
    audio_opt = []
    for text in texts:
        if only_punc(text):
            continue

        # 分割文本，找到所有 (pause:...) 的位置
        parts = re.split(r'\(pause:[0-9.]+\)', text)
        pauses = re.findall(r'\(pause:[0-9.]+\)', text)
        for i, part in enumerate(parts):
            if i > 0:
                pause_time = float(pauses[i-1][7:-1])  # 提取停顿时间
                print(f"加入{pause_time}s停顿")
                stop_wav = np.zeros(
                    int(hps.data.sampling_rate * pause_time),  # 根据停顿时间生成静音
                    dtype=np.float16 if is_half else np.float32,
                )
                audio_opt.append(stop_wav)  # 插入静音片段

            # 处理当前部分的文本
            if part.strip() == "":
                continue  # 如果是空字符串，跳过

            if part[-1] not in splits:  # 检查文本是否以标点符号结尾
                part += "。" if text_languageuage != "en" else "."

            phones2, bert2, norm_text2 = get_phones_and_bert(part, text_languageuage, version)
            bert = torch.cat([bert1, bert2], 1)
            all_phoneme_ids = torch.LongTensor(phones1 + phones2).to(device).unsqueeze(0)
            bert = bert.to(device).unsqueeze(0)
            all_phoneme_len = torch.tensor([all_phoneme_ids.shape[-1]]).to(device)

            with torch.no_grad():
                pred_semantic, idx = t2s_model.model.infer_panel(
                    all_phoneme_ids,
                    all_phoneme_len,
                    prompt,
                    bert,
                    top_k=top_k,
                    top_p=top_p,
                    temperature=temperature,
                    early_stop_num=hz * max_sec,
                )
                pred_semantic = pred_semantic[:, -idx:].unsqueeze(0)

            audio = (
                vq_model.decode(
                    pred_semantic,
                    torch.LongTensor(phones2).to(device).unsqueeze(0),
                    refers,
                    speed=speed,
                )
                .detach()
                .cpu()
                .numpy()[0, 0]
            )
            max_audio = np.abs(audio).max()
            if max_audio > 1:
                audio /= max_audio
            audio_opt.append(audio)  # 添加生成的音频片段

        audio_opt.append(zero_wav)  # 在每行文本的末尾添加静音片段
        t4 = ttime()
        if is_int32:
            audio_bytes = pack_audio(audio_bytes,(np.concatenate(audio_opt, 0) * 2147483647).astype(np.int32),hps.data.sampling_rate)
        else:
            audio_bytes = pack_audio(audio_bytes,(np.concatenate(audio_opt, 0) * 32768).astype(np.int16),hps.data.sampling_rate)
    # logger.info("%.3f\t%.3f\t%.3f\t%.3f" % (t1 - t0, t2 - t1, t3 - t2, t4 - t3))
        if stream_mode == "normal":
            audio_bytes, audio_chunk = read_clean_buffer(audio_bytes)
            yield audio_chunk
    
    if not stream_mode == "normal": 
        if media_type == "wav":
            audio_bytes = pack_wav(audio_bytes,hps.data.sampling_rate)
        yield audio_bytes.getvalue()

**Comment by originwufish** [comment]
api.py可以在get_tts_wav中插入代码实现，api_v2.py暂时还不知道
`        # 分割文本，找到所有 (pause:...) 的位置
        parts = re.split(r'\(pause:[0-9.]+\)', text)
        pauses = re.findall(r'\(pause:[0-9.]+\)', text)
        for i, part in enumerate(parts):
            if i > 0:
                pause_time = float(pauses[i-1][7:-1])  # 提取停顿时间
                print(f"加入{pause_time}s停顿")
                stop_wav = np.zeros(
                    int(hps.data.sampling_rate * pause_time),  # 根据停顿时间生成静音
                    dtype=np.float16 if is_half else np.float32,
                )
                audio_opt.append(stop_wav)  # 插入静音片段

            # 处理当前部分的文本
            if part.strip() == "":
                continue  # 如果是空字符串，跳过

            if part[-1] not in splits:  # 检查文本是否以标点符号结尾
                part += "。" if text_languageuage != "en" else "."`

**Comment by originwufish** [comment]
api.py可以在get_tts_wav中插入代码实现，api_v2.py暂时还不知道
`parts = re.split(r'(pause:[0-9.]+)', text)
pauses = re.findall(r'(pause:[0-9.]+)', text)
for i, part in enumerate(parts):
if i > 0:
pause_time = float(pauses[i-1][7:-1]) # 提取停顿时间
print(f"加入{pause_time}s停顿")
stop_wav = np.zeros(
int(hps.data.sampling_rate * pause_time), # 根据停顿时间生成静音
dtype=np.float16 if is_half else np.float32,
)
audio_opt.append(stop_wav) # 插入静音片段

        # 处理当前部分的文本
        if part.strip() == "":
            continue  # 如果是空字符串，跳过

        if part[-1] not in splits:  # 检查文本是否以标点符号结尾
            part += "。" if text_languageuage != "en" else "." `

**Comment by originwufish** [comment]
api.py可以在get_tts_wav中插入代码实现，api_v2.py暂时还不知道
 # 分割文本，找到所有 (pause:...) 的位置
parts = re.split(r'(pause:[0-9.]+)', text)
pauses = re.findall(r'(pause:[0-9.]+)', text)
for i, part in enumerate(parts):
if i > 0:
pause_time = float(pauses[i-1][7:-1]) # 提取停顿时间
print(f"加入{pause_time}s停顿")
stop_wav = np.zeros(
int(hps.data.sampling_rate * pause_time), # 根据停顿时间生成静音
dtype=np.float16 if is_half else np.float32,
)
audio_opt.append(stop_wav) # 插入静音片段

        # 处理当前部分的文本
        if part.strip() == "":
            continue  # 如果是空字符串，跳过

        if part[-1] not in splits:  # 检查文本是否以标点符号结尾
            part += "。" if text_languageuage != "en" else "."

**Comment by RVC-Boss** [comment]
语速调整已支持

**Comment by RVC-Boss** [comment]
以及停顿时长
